pkg/common-controller/snapshot_controller.go

```diff
@@ -252,6 +269,12 @@ func (ctrl *csiSnapshotCommonController) processSnapshotWithDeletionTimestamp(sn
 	// content has been found from content cache store
 	content, err := ctrl.getContentFromStore(contentName)
 	if err != nil {
+		ctrl.metricsManager.RecordMetrics(metrics.Operation{
```

Comment:
In general, is it possible to have the `RecordMetrics` call as a defer right after Start? That way we don't need to remember to call RecordMetrics if we add more exit points in the future.

Reply:
I think we could have some sort defer call to perform RecordMetrics, but we would still need to update the snapshot type before the defer is executed. We could initialize a snapshotStatusType and change it before the defer is executed, but we'd still need to remember to update that.

I think a solution could be to have a custom error type that implements in the error interface and requires a snapshotStatusType assigned along with it. This would allow us to emit the correct snapshotStatusType based on the error returned per function. That might be a little involved for the 1.20 code freeze though.

Reply:
Is it possible to start the timer after we have figured out if it's dynamic or preprovisioned? We should already have the api object with the source that tells us this.

I think it's fine to have a field for status that needs to be updated before returning. What about initializing it to error, and only setting it to success in the one case?

Reply:
Refactored this. The repetitive code is mainly from all of the error metrics emitting, so I added defer calls that only emit metrics if `err != nil`. Success metrics are still manually emitted outside of the defer calls, this is because they are mostly deeper in the call stack.

I initialized snapshotStatusType to be `controller-error`, and manually changed to `invalid-request`/etc when needed.